src/distribution/negative_binomial.rs

```diff
@@ -0,0 +1,487 @@
+use crate::distribution::{Discrete, Gamma, Univariate};
+use crate::function::{beta, gamma};
+use crate::statistics::*;
+use crate::{Result, StatsError};
+use rand::distributions::Distribution;
+use rand::Rng;
+use std::f64;
+
+/// Implements the
+/// [NegativeBinomial](http://en.wikipedia.org/wiki/Negative_binomial_distribution)
+/// distribution
+///
+/// # Examples
+///
+/// ```
+/// use statrs::distribution::{NegativeBinomial, Discrete};
+/// use statrs::statistics::Mean;
+/// use statrs::prec::{almost_eq};
+///
+/// let r = NegativeBinomial::new(4.0, 0.5).unwrap();
+/// assert_eq!(r.mean(), 4.0);
+/// assert!(almost_eq(r.pmf(0), 0.0625, 1e-8));
+/// assert!(almost_eq(r.pmf(3), 0.15625, 1e-8));
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct NegativeBinomial {
+    r: f64,
+    p: f64,
+}
+
+impl NegativeBinomial {
+    /// Constructs a new negative binomial distribution
+    /// with a given `p` probability of the number of successes `r`
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if `p` is `NaN`, less than `0.0`,
+    /// greater than `1.0`, or if `r` is `NaN` or less than `0`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use statrs::distribution::NegativeBinomial;
+    ///
+    /// let mut result = NegativeBinomial::new(4.0, 0.5);
+    /// assert!(result.is_ok());
+    ///
+    /// result = NegativeBinomial::new(-0.5, 5.0);
+    /// assert!(result.is_err());
+    /// ```
+    pub fn new(r: f64, p: f64) -> Result<NegativeBinomial> {
+        if p.is_nan() || p < 0.0 || p > 1.0 || r.is_nan() || r < 0.0 {
+            Err(StatsError::BadParams)
+        } else {
+            Ok(NegativeBinomial { p: p, r: r })
+        }
+    }
+
+    /// Returns the probability of success `p` of
+    /// the negative binomial distribution.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use statrs::distribution::NegativeBinomial;
+    ///
+    /// let r = NegativeBinomial::new(5.0, 0.5).unwrap();
+    /// assert_eq!(r.p(), 0.5);
+    /// ```
+    pub fn p(&self) -> f64 {
+        self.p
+    }
+
+    /// Returns the number `r` of success of this negative
+    /// binomial distribution
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use statrs::distribution::NegativeBinomial;
+    ///
+    /// let r = NegativeBinomial::new(5.0, 0.5).unwrap();
+    /// assert_eq!(r.r(), 5.0);
+    /// ```
+    pub fn r(&self) -> f64 {
+        self.r
+    }
+}
+
+impl Distribution<u64> for NegativeBinomial {
+    fn sample<R: Rng + ?Sized>(&self, r: &mut R) -> u64 {
+        let lambda_gen = Gamma::new(self.r, self.p).unwrap();
+        let lambda = lambda_gen.sample(r);
+
+        let c = (-lambda).exp();
```

Comment:
Ah, this is definitely more elegant. However, Poisson returns an f64 instead of an u64 - is there a reason for this? I can floor the value but it kind of seems like an error in Poisson although fixing it would be a breaking change. Is there a reason for this?

Reply:
Sampling the Poisson distributions typically requires evaluating continuous functions, so `f64` has to be used internally.